core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceResponseConverterTest.java

```diff
@@ -364,46 +401,144 @@ public HttpResult<?> wildcard() {
                 public <T> HttpResult<T> generic() {
                     return (HttpResult<T>) HttpResult.of(ImmutableList.of("a", "b"));
                 }
+            });
 
-                @Get("/header")
-                @AdditionalHeader(name = "header_name_1", value = "header_value_1")
-                @AdditionalHeader(name = "header_name_2", value = "header_value_2")
-                @AdditionalHeader(name = "header_name_1", value = "header_value_3")
-                public void header() {}
+            sb.annotatedService("/custom-response/response-entity", new Object() {
+                @Get("/expect-specified-no-content")
+                @StatusCode(204)
+                public ResponseEntity<Void> expectSpecifiedNoContent() {
+                    // Will send '204 No Content' because it is specified with @StatusCode annotation.
+                    return null;
+                }
 
-                @Get("/header-overwrite")
-                @AdditionalHeader(name = "header_name_1", value = "header_value_changed")
-                public HttpResponse headerOverwrite() {
-                    return HttpResponse.of(ResponseHeaders.of(HttpStatus.OK,
-                                                              HttpHeaderNames.of("header_name_1"),
-                                                              "header_value_unchanged"));
+                @Get("/expect-not-modified")
+                @StatusCode(204)
+                public ResponseEntity<Void> expectNotModified() {
+                    // Will send '304 Not Modified' because ResponseEntity overrides the @StatusCode
+                    // annotation.
+                    return ResponseEntity.of(ResponseHeaders.of(HttpStatus.NOT_MODIFIED));
+                }
+
+                @Get("/expect-unauthorized")
+                public ResponseEntity<HttpResponse> expectUnauthorized() {
+                    // Will send '401 Unauthorized' because the content of ResponseEntity is HttpResponse.
+                    return ResponseEntity.of(
+                            ResponseHeaders.of(HttpStatus.OK),
+                            HttpResponse.of(HttpStatus.UNAUTHORIZED));
+                }
```

Comment:
Should we raise an exception to avoid this confusing behavior? What do you think?

Reply:
There is also warning, but I think the exception is more kind.
If we do, I think we could throw the exception here with the warning.
https://github.com/line/armeria/blob/a13d582d3aa01e2a8282438140d62f27026e73dd/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedService.java#L210-L213